Add ability to select a reason when closing an issue

Fs00 opened the pull request:
This PR implements the ability to specify a reason (_completed_ or _not planned_) when closing an issue by tapping the "Close" action in the app bar.
In #1223 I initially thought that this couldn't be done with the REST API, but luckily I was wrong.

![close_as_new](https://github.com/user-attachments/assets/2b5e902c-860e-4852-bc8b-68b8d5c3d0b1)

I intentionally kept the close reason descriptions short (almost the same as the GH web UI) so that they fit on a single line even on smaller screens. 

One minor feature that this PR lacks compared to the GH web UI is the ability to change the close reason of an issue without first reopening it. I didn't bother to add it, since I think it will rarely be useful in practice.
In case we really want to have it, we could simply add a "Close as not planned"/"Close as completed" action in the dropdown menu when the issue has already been closed.

maniac103 commented on app/src/main/res/values/strings.xml:
```diff
@@ -432,8 +432,13 @@
     <string name="issue_no_description">No description provided.</string>
-    <string name="close_issue_confirm">Do you really want to close this issue?</string>
     <string name="reopen_issue_confirm">Do you really want to reopen this issue?</string>
+    <string name="choose_reason">Choose a reason</string>
```
How about `Close issue as\u2026`?

Fs00 replied:
I agree that it suits better. I've pushed the change and updated the screenshot in the description.

maniac103 commented:
Thinking about it, maybe there's potential to consolidate with [this code](https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/activities/IssueEditActivity.java#L827) used in a similar dialog? It uses a dedicated adapter instead of the (creative, but still ;-) ) string builder hack.
Maybe refactor into a `TwoLineSelectionDialog` or something? Or share the adapter at least?

Fs00 commented:
Hmm... I think that having a single dialog for both use-cases will not help, since the user interaction is different between the two (checking a radio button and confirming _vs_ just tapping on an item in the list).
Also, having a common adapter doesn't seem worth it because the item layout is different in this case, since we have the radio button in addition to just the title and description.
If you'd like to avoid the code messing around with the `SpannableStringBuilder`, I could try to use a dedicated adapter + layout for the items (something similar to the solutions proposed in [this SO thread](https://stackoverflow.com/questions/59696021/how-to-add-title-and-subtitle-in-android-radiobutton)), although I think that the implementation will be a bit more complex overall.

maniac103 commented:
> Hmm... I think that having a single dialog for both use-cases will not help, since the user interaction is different between the two (checking a radio button and confirming vs just tapping on an item in the list).

Why not make both confirm on item tap? ;-) Technically, the current solution is 'confirm on item tap' as well, even though the item currently is a dialog button.

Fs00 commented:
Could be an interesting idea, I'll try and see how it looks

Fs00 commented:
I've just updated the implementation to use the same layout and behavior as the issue template selection dialog (screenshot in the description updated), which allowed me to reuse the same adapter for both dialogs.
While I was at it, I've discovered and fixed two minor bugs in the issue template dialog:
- the order of issue templates as returned by the API was not deterministic, so we now sort them by name
- tapping the Cancel button didn't fire the `onCancel` callback as the previous code assumed, and therefore in that case the editor would have remained stuck on the "Loading issue templates" message

To make my life easier, I have also enabled support for Java 17 features (which include records, switch expressions, text blocks and some more) and desugaring for Java 8/11 library features (for using streams and `Comparator.comparing`).
